File: diagonal-traverse/diagonal-traverse.java
class Solution {
    public int[] findDiagonalOrder(int[][] matrix) {
        if (matrix.length==0) return new int[0];
        int m = matrix.length;
        int n = matrix[0].length;
        int[] result = new int[m*n];
        int index = 0;
        
        int row = 0, col = 0, diagonal = 1;
        
        while ((row<=m-1 && col<=n-1)) {
            
            if (diagonal == 0) {
                //leftwards
                int r,c;
                for (r = row,c = col; r<m && c>=0; r++,c--) {
                        result[index++] = matrix[r][c];
                }
                row = r-1; col = c+1;
                if (row+1<m) row++;
                else col++;
                diagonal = 1;
            }
            
            else {
                //rightwards
                int r,c;
                for ( r = row, c = col; r>=0 && c<n; r--,c++) {
                        result[index++] = matrix[r][c];
                }
                row = r+1; col = c-1;
                if (col+1<n) col++;
                else row++;
                diagonal = 0;
            }
        }
        return result;
    }
}

File: 498-diagonal-traverse/diagonal-traverse.cpp
class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& mat) {
        vector<int> arr;
        int i=0,j=0;
        while(arr.size() < mat.size()*mat[0].size())
        {
            arr.push_back(mat[i][j]);
            if((i+j)%2==0)
            {
                if(j==mat[0].size()-1){i++;}
                else if(i==0){j++;}
                else{i--;j++;}
            }
            else
            {
                if(i==mat.size()-1){j++;}
                else if(j==0){i++;}
                else{i++;j--;}
            }
        }
        return arr;
    }
};